backend/src/main/java/com/google/sps/servlets/AddParticipantServlet.java

```diff
@@ -0,0 +1,151 @@
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.sps.data.Match;
+import com.google.sps.data.Participant;
+import com.google.sps.FindMatchQuery;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+* Servlet that returns some example content.
+*/
+@WebServlet("/add-participant")
+public class AddParticipantServlet extends HttpServlet {
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Request parameter values
+    // TODO: Error check input values (before allowing to submit)
+    UserService userService = UserServiceFactory.getUserService();
+    String email = userService.getCurrentUser().getEmail();
+    String ldap = email.split("@")[0];
+    long timeAvailableUntil = convertToLong(request.getParameter("timeAvailableUntil"));
+    String timezone = request.getParameter("timezone");
+    int duration = convertToInt(request.getParameter("duration"));
+    long timestamp = System.currentTimeMillis();
+
+    // id is irrelevant, only relevant when getting from datastore
+    Participant newParticipant = new Participant(/* id= */ -1L, ldap, timeAvailableUntil, timezone, duration, timestamp);
+    
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    // Find immediate match if possible
+    FindMatchQuery query = new FindMatchQuery();
+    Match match = query.findMatch(getParticipants(datastore), newParticipant);
+
+    // Match found, add to datastore, delete matched participants from datastore
+    if (match != null) {
+      addMatchToDatastore(match, datastore);
+      deleteSecondParticipantFromDatastore(match.getSecondParticipant(), datastore);
+    } else {
+      // Match not found, insert participant entity into datastore
+      Entity participantEntity = new Entity("Participant");
+      participantEntity.setProperty("ldap", ldap);
+      participantEntity.setProperty("timeAvailableUntil", timeAvailableUntil);
+      participantEntity.setProperty("timezone", timezone);
+      participantEntity.setProperty("duration", duration);
+      participantEntity.setProperty("timestamp", timestamp);
+      datastore.put(participantEntity);
+    }
+
+    // Redirect back to the HTML page
+    response.sendRedirect("/index.html");
+  }
+
+  /** Return list of current participants from datastore*/
+  private List<Participant> getParticipants(DatastoreService datastore) {
+
+    // Create and sort participant queries by time
+    Query query = new Query("Participant").addSort("timestamp", SortDirection.DESCENDING);
```

Comment:
is there any way to narrow this query down to participants available within a desired time range? For a sufficiently large set of participants, loading the entire table into memory could be expensive, even prohibitively so.

Reply:
The only way I can think of narrowing this query down is by moving determination of compatibility based on timeAvailableUntil and duration into this method instead of in FindMatchQuery. Should I still do that? It doesn't really make sense when the matching alg only involves time compatibility but seems to make more sense once I add in personal fields (role, product area, etc.)